### PR TITLE
[Transformation] If redundant lines removal

### PR DESCRIPTION
**Pull request by nbirillo**

### CODE REVIEW

File: src/main/kotlin/org/jetbrains/research/ml/ast/transformations/ifRedundantLinesRemoval/IfRedundantLinesRemover.kt [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+import org.jetbrains.research.ml.ast.transformations.createBinaryOperandList
+import org.jetbrains.research.ml.ast.transformations.createExpressionStatement
+import org.jetbrains.research.ml.ast.transformations.createIfPart
+import org.jetbrains.research.ml.ast.transformations.createIfPartFromIfPart
+import org.jetbrains.research.ml.ast.transformations.createPrefixExpression
+import org.jetbrains.research.ml.ast.transformations.safePerformCommand
+import org.jetbrains.research.ml.ast.transformations.safePerformCommandWithResult
+import kotlin.test.fail
+
+class IfRedundantLinesRemover(
```
Comment (elena-lyulina) [inline src/main/kotlin/org/jetbrains/research/ml/ast/transformations/ifRedundantLinesRemoval/IfRedundantLinesRemover.kt:18]: This class is so big, can we split it somehow? For example, extract all functions about common suffix and prefix? Or anything else?

File: src/test/resources/org/jetbrains/research/ml/ast/transformations/ifRedundantLinesRemoval/data/pure/in_4_if_elif_else.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+s = input()
+
+if s:
```
Comment (elena-lyulina) [inline src/test/resources/org/jetbrains/research/ml/ast/transformations/ifRedundantLinesRemoval/data/pure/in_4_if_elif_else.py:3]: A little bit strange that you use `if s:` - `elif s:`. Maybe it does not affect the transformation, but what if it does?

**Review by elena-lyulina — CHANGES_REQUESTED** [review]
Looks great, thanks!

File: docs/transformations/IfRedundantLinesRemovalTransformation.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+2. Calculate the length of the common suffix and prefix
+
+3. Move before and after of the operator the common suffix and prefix 
+
+4. Remove the common parts from the `if` operator
+
+5. Validate the new `if` operator
+
+
+## Exceptions
```
[the remark this answers is not in the document]
Comment (nbirillo) [reply docs/transformations/IfRedundantLinesRemovalTransformation.md:58]: Maybe we should compare PSI by another metric, but the first exception is ok for our task I think. Anyway, I would change it in the future, not now

File: src/main/kotlin/org/jetbrains/research/ml/ast/transformations/ifRedundantLinesRemoval/IfRedundantLinesRemover.kt [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+    private fun simplifySubStatementsDelayed(element: PsiElement): () -> Unit {
+        val simplifySubStatements = element.children.map { simplifyAllDelayed(it) }
+        return { simplifySubStatements.forEach { it() } }
+    }
+
+    private data class StatementRange(val first: PyStatement, val last: PyStatement) {
+        fun move(
+            anchor: PsiElement,
+            before: Boolean,
```
[the remark this answers is not in the document]
Comment (elena-lyulina) [reply src/main/kotlin/org/jetbrains/research/ml/ast/transformations/ifRedundantLinesRemoval/IfRedundantLinesRemover.kt]: Please, rename isBefore -> doBefore on line 220

File: src/main/kotlin/org/jetbrains/research/ml/ast/transformations/ifRedundantLinesRemoval/IfRedundantLinesRemover.kt (hunk 1 above)
Comment (elena-lyulina) [reply src/main/kotlin/org/jetbrains/research/ml/ast/transformations/ifRedundantLinesRemoval/IfRedundantLinesRemover.kt]: Sorry, don't see any changes, so what do you think about splitting the class?

**Review by elena-lyulina — CHANGES_REQUESTED** [review]
Sorry, I noticed you marked some comments as resolved but you didn't complete fixes. Could you please check it one more time?

File: src/main/kotlin/org/jetbrains/research/ml/ast/transformations/ifRedundantLinesRemoval/IfRedundantLinesRemover.kt (hunk 1 above)
Comment (nbirillo) [reply src/main/kotlin/org/jetbrains/research/ml/ast/transformations/ifRedundantLinesRemoval/IfRedundantLinesRemover.kt]: I made two inner classes `StatementRemover` and `StatementSimplifier` to separate some logic

**Review by elena-lyulina — APPROVED** [review]
Great!

**Pull request merged**